shared/team-building/container.js

```diff
@@ -0,0 +1,198 @@
+// @flow
+import logger from '../logger'
+import React from 'react'
+import * as I from 'immutable'
+import TeamBuilding from '.'
+import * as TeamBuildingGen from '../actions/team-building-gen'
+import {type TypedState, compose, connect, setDisplayName, createSelector} from '../util/container'
+import {PopupDialogHoc} from '../common-adapters'
+import {parseUserId} from '../util/platforms'
+import {followStateHelperWithId} from '../constants/team-building'
+import type {ServiceIdWithContact, User} from '../constants/types/team-building'
+
+// TODO
+// * there's a lot of render thrashing going on. using keyboard arrows is kinda slow becuase of it.
+// * Limit the highlight index to the max lenght of the list
+
+type OwnProps = {
+  // Supplied by StateComponent
+  searchString: ?string,
+  selectedService: ServiceIdWithContact,
+  highlightedIndex: ?number,
+  clearTextTrigger: number,
+  onChangeText: (newText: string) => void,
+  onChangeService: (newService: ServiceIdWithContact) => void,
+  onDownArrowKeyDown: () => void,
+  onUpArrowKeyDown: () => void,
+  incClearTextTrigger: () => void,
+}
+
+type LocalState = {
+  searchString: ?string,
+  selectedService: ServiceIdWithContact,
+  highlightedIndex: ?number,
+  clearTextTrigger: number,
+}
+
+const initialState: LocalState = {
+  searchString: null,
+  selectedService: 'keybase',
+  highlightedIndex: 0,
+  clearTextTrigger: 0,
+}
+
+const searchResultsSelector = (state: TypedState, ownProps: OwnProps) =>
+  state.chat2.teamBuildingSearchResults.get(I.List([ownProps.searchString, ownProps.selectedService]))
+
+const searchResultsPropSelector = createSelector(
+  searchResultsSelector,
+  (state: TypedState) => state.chat2.teamBuildingTeamSoFar,
+  (state: TypedState) => state.config.username,
+  (state: TypedState) => state.config.following,
+  (
+    searchResults: ?Array<User>,
+    teamSoFar: I.Set<User>,
+    myUsername: string,
+    followingState: I.Set<string>
+  ) => {
+    return (searchResults || []).map(info => ({
+      userId: info.id,
+      username: info.id.split('@')[0],
+      services: info.serviceMap,
+      prettyName: info.prettyName,
+      followingState: followStateHelperWithId(myUsername, followingState, info.id),
+      inTeam: teamSoFar.has(info),
+    }))
+  }
+)
+
+const teamSoFarPropSelector = createSelector(
+  (state: TypedState) => state.chat2.teamBuildingTeamSoFar,
+  (teamSoFar: I.Set<User>) =>
+    teamSoFar.toArray().map(userInfo => {
+      const {username, serviceId} = parseUserId(userInfo.id)
+      return {
+        userId: userInfo.id,
+        prettyName: userInfo.prettyName,
+        service: serviceId,
+        username,
+      }
+    })
+)
+
+const mapStateToProps = (state: TypedState, ownProps: OwnProps) => {
+  return {
+    searchResults: searchResultsSelector(state, ownProps),
+    searchResultsProp: searchResultsPropSelector(state, ownProps),
```

Comment:
I'm not happy with how I have `searchResults` and `searchResultsProp`. I currently need them both since the Prop version is what gets passed to the component, but the plain version is used when you add a user. So we can return the same `User` obj the store gave us.